forest_benchmarking/qubit_spectroscopy.py

```diff
@@ -50,58 +64,144 @@ def generate_t1_experiments(qubits: Union[int, List[int]],
     :param stop_time: The maximum decay time to measure at.
     :param n_shots: The number of shots to average over for each data point.
     :param num_points: The number of points for each t1 curve.
-    :return: list of tuples in the form: (time, t1 program with decay of that time)
+    :return: A dataframe with columns: time, t1 program
     """
     start_time = 0
     time_and_programs = []
-    for t in np.linspace(start_time, stop_time, num_points):
+
+    for t in np.linspace(start_time, stop_time, n_points):
         t = round(t, 7)  # try to keep time on 100ns boundaries
-        time_and_programs.append((t, generate_single_t1_experiment(qubits, t, n_shots)))
-    return time_and_programs
+        time_and_programs.append({
+            'times': t,
+            'programs': generate_single_t1_experiment(qubits, t, n_shots)
+        })
+    return pd.DataFrame(time_and_programs)
 
 
-def run_t1(qc: QuantumComputer,
-           qubits: Union[int, List[int]],
-           stop_time: float,
-           n_shots: int = 1000,
-           num_points: int = 15,
-           filename: str = None) -> pd.DataFrame:
+def acquire_data_t1(qc: QuantumComputer,
+                    t1_experiment: pd.DataFrame,
+                    ) -> pd.DataFrame:
     """
     Execute experiments to measure the t1 decay time of 1 or more qubits.
 
-    :param qc: The QuantumComputer to run the experiment on.
-    :param qubits: Which qubits to measure.
-    :param stop_time: The maximum decay time to measure at.
-    :param n_shots: The number of shots to average over for each data point.
-    :param num_points: The number of points for each t1 curve.
-    :param filename: The name of the file to write JSON-serialized results to.
+    :param qc: The QuantumComputer to run the experiment on
+    :param t1_experiment: A pandas DataFrame with columns: time, t1 program
+    :return: pandas DataFrame
     """
     results = []
-    for t, program in generate_t1_experiments(qubits, stop_time, n_shots, num_points):
+
+    for index, row in t1_experiment.iterrows():
+        t = row['times']
+        program = row['programs']
+
         executable = qc.compiler.native_quil_to_executable(program)
         bitstrings = qc.run(executable)
 
+        qubits = list(program.get_qubits())
         for i in range(len(qubits)):
             avg = np.mean(bitstrings[:, i])
             results.append({
                 'qubit': qubits[i],
                 'time': t,
                 'n_bitstrings': len(bitstrings),
                 'avg': float(avg),
+                'programs': program,
```

Comment:
We could populate the qubit column in generate_experiments and start off the dataframe with the per-qubit separation, where each program only acts on one qubit. Here, we would then optionally group programs by time and run groups simultaneously, but separately populate each row with results only for that qubit. This will need to match whatever happens with RPE, so it is fine to leave like this for now.

Reply:
Ok that sounds fair. I could do that and the other change when we deicide what is happening with RPE.